drivers/flash/flash_rpi_pico.c

```diff
@@ -0,0 +1,307 @@
+
+#define PAGE_SIZE 256
+#define SECTOR_SIZE DT_PROP(DT_CHILD(DT_NODELABEL(flash_controller), flash_0), erase_block_size)
+#define ERASE_VALUE 0xff
+#define FLASH_SIZE KB(CONFIG_FLASH_SIZE)
+
+static const struct flash_parameters flash_rpi_parameters = {
+	.write_block_size = 1,
+	.erase_value = ERASE_VALUE,
+};
+
+/**
+ * Low level flash functions are based on:
+ * github.com/raspberrypi/pico-bootrom/blob/master/bootrom/program_flash_generic.c
+ * and
+ * github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_flash/flash.c
+ */
+
+#define FLASHCMD_PAGE_PROGRAM 0x02
+#define FLASHCMD_READ_STATUS 0x05
+#define FLASHCMD_WRITE_ENABLE 0x06
+#define BOOT2_SIZE_WORDS 64
+
+enum outover {
+	OUTOVER_NORMAL = 0,
+	OUTOVER_INVERT,
+	OUTOVER_LOW,
+	OUTOVER_HIGH
+};
+
+static ssi_hw_t *const ssi = (ssi_hw_t *) XIP_SSI_BASE;
+static uint32_t boot2_copyout[BOOT2_SIZE_WORDS];
+static bool boot2_copyout_valid;
+
+static void __no_inline_not_in_flash_func(flash_init_boot2_copyout)(void) {
+	if (boot2_copyout_valid)
+		return;
+	for (int i = 0; i < BOOT2_SIZE_WORDS; ++i)
+		boot2_copyout[i] = ((uint32_t *)XIP_BASE)[i];
+	__compiler_memory_barrier();
+	boot2_copyout_valid = true;
+}
+
+static void __no_inline_not_in_flash_func(flash_enable_xip_via_boot2)(void) {
+	((void (*)(void))boot2_copyout+1)();
+}
+
+void __no_inline_not_in_flash_func(flash_cs_force)(enum outover over) {
+	io_rw_32 *reg = (io_rw_32 *) (IO_QSPI_BASE + IO_QSPI_GPIO_QSPI_SS_CTRL_OFFSET);
+	*reg = (*reg & ~IO_QSPI_GPIO_QSPI_SS_CTRL_OUTOVER_BITS)
+		| (over << IO_QSPI_GPIO_QSPI_SS_CTRL_OUTOVER_LSB);
+	(void) *reg;
+}
+
+int __no_inline_not_in_flash_func(flash_was_aborted)() {
+	return *(io_rw_32 *) (IO_QSPI_BASE + IO_QSPI_GPIO_QSPI_SD1_CTRL_OFFSET)
+		   & IO_QSPI_GPIO_QSPI_SD1_CTRL_INOVER_BITS;
+}
+
+void __no_inline_not_in_flash_func(flash_put_get)(const uint8_t *tx, uint8_t *rx, size_t count,
+						size_t rx_skip) {
+	const uint max_in_flight = 16 - 2;
+	size_t tx_count = count;
+	size_t rx_count = count;
+	bool did_something;
+	uint32_t tx_level;
+	uint32_t rx_level;
+	uint8_t rxbyte;
+
+	while (tx_count || rx_skip || rx_count) {
+		tx_level = ssi_hw->txflr;
+		rx_level = ssi_hw->rxflr;
+		did_something = false;
+		if (tx_count && tx_level + rx_level < max_in_flight) {
+			ssi->dr0 = (uint32_t) (tx ? *tx++ : 0);
+			--tx_count;
+			did_something = true;
+		}
+		if (rx_level) {
+			rxbyte = ssi->dr0;
+			did_something = true;
+			if (rx_skip) {
+				--rx_skip;
+			} else {
+				if (rx)
+					*rx++ = rxbyte;
+				--rx_count;
+			}
+		}
+
+		if (!did_something && __builtin_expect(flash_was_aborted(), 0))
+			break;
+	}
+	flash_cs_force(OUTOVER_HIGH);
+}
+
+void __no_inline_not_in_flash_func(flash_put_get_wrapper)(uint8_t cmd, const uint8_t *tx,
+					uint8_t *rx, size_t count) {
+	flash_cs_force(OUTOVER_LOW);
+	ssi->dr0 = cmd;
+	flash_put_get(tx, rx, count, 1);
+}
+
+void __no_inline_not_in_flash_func(flash_put_cmd_addr)(uint8_t cmd, uint32_t addr) {
```

Comment:
This function is [inlined](https://github.com/raspberrypi/pico-bootrom/blob/master/bootrom/program_flash_generic.c#L160) for timing reasons. Why the __noinline here?